### PR TITLE
devices: support for MX Vertical mouse

### LINKED ISSUE #573: MX Vertical
Hi, are there any plans for MX Vertical support?

### PR DESCRIPTION
**Pull request by Vtec234**
Closes #573

Implementation status:

- [x] descriptor and DPI button ID for MX Vertical mouse
- [x] full support for `REPROG_CONTROLS_V4`
- [x] backend interface for diverting and remapping controls
- [x] initial support for diverted press/raw XY notifications
- [x] controlling DPI via sliding on the MX Vertical

### CODE REVIEW

**Comment by pfps** [comment]
Would it be useful to separate out the different parts of this change into different commits?

**Comment by Vtec234** [comment]
I split it as much as possible, hopefully will be a bit easier to review now (although the `REPROG_CONTROLS_V4` commit is still quite large). I should be done with all the functionality soon, so feel free to wait until then with looking at this.

**Comment by pfps** [comment]
Excellent.  Yes, I realize that the key/button reprogramming bit is large.   Fortunately I have several devices that I think use this feature so I should be able to check out how it works.

**Comment by FFY00** [comment]
That is good but I would appreciate if you could drop the style changes from the patch. If you think they should still be applied please open a new PR to discuss.

**Comment by Vtec234** [comment]
> That is good but I would appreciate if you could drop the style changes from the patch. If you think they should still be applied please open a new PR to discuss.

Yeah, sorry. I undid the formatting so the diff should be much clearer now. My suggestion - run [black](https://github.com/psf/black) on the entire codebase and never think about style again, but you're right in that it should be done separately.

**Comment by pfps** [comment]
Summary:
I thought that this would have key reprogramming, which I could check.  The major functional changes are for the DPI setting on this mouse, which I don't have and thus I can't check the changes.
The `solaar show` output needs a slight fix.
It would be better to not introduce and then remove the style changes.

A bit more on each commit:
**note: slowness**
	- Just adding a comment
**misc: undo stylistic changes and use translation**
	- This mostly (?) undoes the style change in the second commit.
	- It would be better to eliminate the style changes entirely.
**device: support DPI sliding on MX Vertical mouse**
	- I can't test this out.  Who has this mouse and can test it.
**misc: typo**
	- It would be better to amend the place where this problem was introduced
**misc: handle unexpected device numbers graciously**
	- OK (minor change)
	- By the way, what is the f before strings for?  Is it different from _?
**misc: variable name fixes**
	- OK (minor change)
**backend: allow device objects to handle notifications**
	- OK
**device: descriptor and control ID for MX Vertical mouse**
	- OK
**backend: initial support for REPROG_CONTROLS_V4 notifications**
	- OK - needs debugging wrapped in check for flag being on but that is in a later commit
**backend: support REPROG_CONTROLS_V4 and divertable/reprogrammable controls**
	- Includes style changes - they should probably be removed to show actual changes.
	- Actual changes that I can see: (Is this all of what is going on here?)
	1/classes for reprogrammable keys
	2/different output from solaar show
	 - Has extra line in output after "mapping flags:" - needs fixing
	- Otherwise works correctly for my device with this feature.
**misc: report exceptions in more detail**
	OK (minor change)

**Comment by FFY00** [comment]
The codestyle undo commit should be squashed with the second one. 

This patches need testing before merging.

**Comment by FFY00** [comment]
ping #573

**Comment by Vtec234** [comment]
Hey, thanks a lot for reviewing the changes! Again sorry about the codestyle stuff, it slipped in because I had autoformatting on. I'm proposing that it's not really a problem, because:
- for review, you can look at `hidpp20.py` in the overall all-commits diff view, and that shows exactly the real changes I made
- for a nicer history, I would just squash-and-merge this PR

If you **really** want to preserve all commits, I can cleanup the history, but since it's as low on my list of priorities as it can possibly be, it might take a while before I get around to it.

> I thought that this would have key reprogramming, which I could check.

It does, and you should be able to test it! I didn't add any frontend for it, but I tried to fully document this functionality in the `ReprogrammableKeyV4` interface. Testing should be as easy as chucking something like `dev.keys[nkey].remap(target_control_id)` into your favourite spot, e.g. `_print_device` in `show.py` (just for testing, ofc).

> device: support DPI sliding on MX Vertical mouse
> I can't test this out. Who has this mouse and can test it.

I've been using this feature since I made the PR and it works great. The scaling is a bit janky and a visual indicate of the DPI level would be useful, but the core functionality is there. I agree it would be great if someone else who also has this mouse could test the PR.

> By the way, what is the f before strings for? Is it different from _?

https://docs.python.org/3/reference/lexical_analysis.html#f-strings

>  - Has extra line in output after "mapping flags:" - needs fixing

This is completely intentional, because otherwise the lines get too long. Here's the output for a diverted key:
```
         5: MX VERTICAL DPI           , default action: unknown:00D2                => mapped to MX VERTICAL DPI
             mapping flags: diverted, rawXY diverted
             mouse, reprogrammable, divertable, persistently divertable, pos:0, group:2, group mask:g1, g2
```

As you can see, the lines would be quite long if I joined them. I could do it if you want, but IMO it's nicer this way.

**Comment by fcastilloec** [comment]
I have an MX Vertical mouse, and I tried this PR but the DPI button still doesn't work. Here's an output of what I got by running `solaar -d` from this branch/PR:
```
17:32:15,338     INFO [MainThread] root: language en_US (UTF-8), translations path None
17:32:15,407     INFO [MainThread] solaar.upower: connected to system dbus, watching for suspend/resume events
17:32:15,430     INFO [MainThread] solaar.ui.notify: starting desktop notifications
17:32:15,490     INFO [MainThread] solaar.listener: starting receiver listening threads
17:32:15,491     INFO [MainThread] solaar.listener: receiver event add DeviceInfo(path='/dev/hidraw0', vendor_id='046d', product_id='c52b', serial='', release=b'2407', manufacturer=b'Logitech', product=b'USB Receiver', interface=2, driver='logitech-djreceiver')
17:32:15,499     INFO [ReceiverListener:hidraw0] logitech_receiver.listener: started with <UnifyingReceiver(/dev/hidraw0,12)> (12)
17:32:15,499     INFO [ReceiverListener:hidraw0] solaar.listener: <UnifyingReceiver(/dev/hidraw0,12)>: notifications listener has started (12)
17:32:15,503     INFO [ReceiverListener:hidraw0] logitech_receiver.receiver: <UnifyingReceiver(/dev/hidraw0,12)>: receiver notifications enabled => ('wireless', 'software present')
17:32:15,507     INFO [ReceiverListener:hidraw0] solaar.listener: status_changed <UnifyingReceiver(/dev/hidraw0,12)>: present, No paired devices. (0) 
17:32:15,508     INFO [ReceiverListener:hidraw0] logitech_receiver.receiver: <UnifyingReceiver(/dev/hidraw0,12)>: found new device 1 (407B)
17:32:15,508     INFO [ReceiverListener:hidraw0] solaar.listener: Notification(1,41,04,327B40) triggered new device <PairedDevice(1,407B,MX Vertical,50173AE6)> (mouse)
17:32:15,509     INFO [ReceiverListener:hidraw0] solaar.listener: status_changed <UnifyingReceiver(/dev/hidraw0,12)>: present, 1 paired device. (0) 
17:32:15,826     INFO [ReceiverListener:hidraw0] solaar.configuration: saved {'0E40:50173AE6': {'_name': 'MX Vertical', 'dpi': 1000, 'lowres-smooth-scroll': False}, '407B:50173AE6': {'_name': 'Wireless Mouse MX Vertical', 'dpi': '1300', 'lowres-smooth-scroll': False, 'mx-vertical-dpi-sliding': True}, '_version': '1.0.1'} to /home/felipe/.config/solaar/config.json
17:32:16,148     INFO [ReceiverListener:hidraw0] solaar.configuration: saved {'0E40:50173AE6': {'_name': 'MX Vertical', 'dpi': 1000, 'lowres-smooth-scroll': False}, '407B:50173AE6': {'_name': 'Wireless Mouse MX Vertical', 'dpi': '1300', 'lowres-smooth-scroll': False, 'mx-vertical-dpi-sliding': True}, '_version': '1.0.1'} to /home/felipe/.config/solaar/config.json
17:32:16,168     INFO [ReceiverListener:hidraw0] solaar.configuration: saved {'0E40:50173AE6': {'_name': 'MX Vertical', 'dpi': 1000, 'lowres-smooth-scroll': False}, '407B:50173AE6': {'_name': 'Wireless Mouse MX Vertical', 'dpi': '1300', 'lowres-smooth-scroll': False, 'mx-vertical-dpi-sliding': True}, '_version': '1.0.1'} to /home/felipe/.config/solaar/config.json
17:32:16,239     INFO [ReceiverListener:hidraw0] solaar.listener: status_changed <PairedDevice(1,407B,MX Vertical,50173AE6)>: paired online, {'LINK ENCRYPTED': True, 'BATTERY LEVEL': 50, 'BATTERY STATUS': NamedInt(0, 'discharging'), 'BATTERY CHARGING': False, 'ERROR': None} (1) 
17:32:16,240     INFO [ReceiverListener:hidraw0] solaar.listener: status_changed <PairedDevice(1,407B,MX Vertical,50173AE6)>: paired online, {'LINK ENCRYPTED': True, 'BATTERY LEVEL': 50, 'BATTERY STATUS': NamedInt(0, 'discharging'), 'BATTERY CHARGING': False, 'ERROR': None} (0) 
17:32:16,240  WARNING [ReceiverListener:hidraw0] logitech_receiver.notifications: <PairedDevice(1,407B,MX Vertical,50173AE6)>: unknown BATTERY Notification(1,08,10,0404C096050000000000000000000000)
```
You can see that `mx-vertical-dpi-sliding` is set to `True` but when pressing the DPI button I don't notice any change. The GUI still shows `1300` as my DPI even after multiple presses.
Running `solaar -dd` on this branch shows this when pressing the DPI button:
```
17:29:44,186    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (12) => r[11 01 0A00 00FD0000000000000000000000000000]
17:29:44,428    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (12) => r[11 01 0A00 00000000000000000000000000000000]
```
In contrast, this is what's shown for Solaar from the Git PPA repository (I think it was last updated 2020-04-05):
```
17:45:08,094    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (12) => r[11 01 0A00 00FD0000000000000000000000000000]
17:45:08,095  WARNING [ReceiverListener:hidraw0] logitech_receiver.notifications: <PairedDevice(1,407B,MX Vertical,50173AE6)>: unrecognized Notification(1,0A,00,00FD0000000000000000000000000000) for feature REPROG CONTROLS V4 (index 0A)
17:45:11,454    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (12) => r[11 01 0A00 00000000000000000000000000000000]
17:45:11,454  WARNING [ReceiverListener:hidraw0] logitech_receiver.notifications: <PairedDevice(1,407B,MX Vertical,50173AE6)>: unrecognized Notification(1,0A,00,00000000000000000000000000000000) for feature REPROG CONTROLS V4 (index 0A)
```

Let me know how I can help to get this working, I'm happy to help.

**Comment by Vtec234** [comment]
@fcastilloec Thanks for having a look! The DPI should *not* change on just pressing the DPI button - the intended usage is to hold said button and then slide the mouse horizontally to adjust the DPI value. `solaar -d` should print some INFO events when you do that - could you please test this?

**Comment by fcastilloec** [comment]
@Vtec234 thanks for letting me know. I can report that it does work but it's not ideal. I can see the messages that the DPI is changing and indeed feel the difference. This is amazing work!!!

The problem I've encountered is: if you move the mouse slowly (~1cm/s) while attempting to switch DPIs, there are no changes at all; it almost feels like it needs acceleration rather than a constant speed to change. Once you move it fast enough then it changes very fast, to extremes (400 to 4,000) and it's very hard to land in a specific value.
It feels that to calculate the switching between values, speed/acceleration and the amount of travel the mouse has on the screen are the main variables. This last part makes it really hard to switch to a higher DPI from a lower value because at 400 DPI, the mouse barely travels and it requires aggressive movements to switch values, which usually just sends the value straight to 4,000 DPI once a change is registered.
I can see on the console that Solaar is registering every single movement the mouse makes when changing DPI but it usually stays at the same value because of the above-mentioned reasons.

I think that some form of GUI/notification would make it a lot easier to switch values and also updating the Solaar screen with the new DPI would be helpful.

Last, what I'm hoping this PR can also implement is having a set of two DPI values that I can switch to by just pressing the button, without any mouse movement. This is a functionality that the mouse has on Windows. Hopefully, it won't be too hard to implement given all this amazing work that has been done already.
Thanks again for the hard work!

**Comment by Vtec234** [comment]
> if you move the mouse slowly (~1cm/s) while attempting to switch DPIs, there are no changes at all; it almost feels like it needs acceleration rather than a constant speed to change. Once you move it fast enough then it changes very fast, to extremes (400 to 4,000) and it's very hard to land in a specific value.

Yep, this is because while the switching should really be done with an accumulator, currently it just branches on whether the mouse moves fast enough because that was easier. I also find this annoying and left a TODO in the code to implement it properly.

> I think that some form of GUI/notification would make it a lot easier to switch values and also updating the Solaar screen with the new DPI would be helpful.

> Last, what I'm hoping this PR can also implement is having a set of two DPI values that I can switch to by just pressing the button, without any mouse movement. This is a functionality that the mouse has on Windows. Hopefully, it won't be too hard to implement given all this amazing work that has been done already.

Agreed on both points. The single-click switching already works by-default on my mouse, but DPI Sliding has to be disabled, and the two values are hardcoded in the firmware rather than configurable. It would be nice for everything to work together and be configurable.

I'll ping you once I implement (some subset of) these features.

**Comment by fcastilloec** [comment]
Thanks again for your work and I'll eagerly wait for an update.

I another note, I know the default on Windows is to slide the mouse horizontally to change the DPI but wouldn't it be easier to use the scrolling wheel for this? I'm not even sure if this is possible and on top of that there's a bigger discussion of changing the default behavior of the mouse. Some users coming from Windows might expect the mouse to work the same everywhere.

**Comment by pfps** [comment]
@FFY00 I haven't had a chance to look closely at this yet.  I'll probably need a weekend with no other stuff that I need to do.   One problem is that I think we should put out a new release soon and an RC release as soon as the distribution stuff is merged in but I don't think this change should go in unless there is better review. 
One option would be to pull out the stuff that is not specific to this mouse.  I can look at that much easier and see whether it is all OK.

**Comment by pfps** [comment]
I'm still trying to free up time to take a closer look at this.

**Comment by FFY00** [comment]
I can make an RC now and then one once this gets merged. Hopefully, we won't have much stuff merged then so we can just wait a few days before making the normal release.

**Comment by pfps** [comment]
I started going through this again.  A big problem is that the main commits also have a lot of style changes in them, which makes it much harder to see what is going on, to the point that I don't have enough time right now to tease out what is what.  Having the three main commits having only the real changes will let me review this sooner.

The first commit is small and useful.  I can review and approve a PR with just this commit.

**Comment by robbat2** [comment]
@fcastilloec @pfps can I suggest that now might be a good time to look at `autopep8` or `pycodestyle`, and introduce auto-consistent formatting? If the codebase in master has it, then all PRs can be auto-formatted to match, and reduce style considerations.

**Comment by pfps** [comment]
@FFY00 Does it sound like a good idea to set up code formatting?   If so, what style should be used and how can it be enforced?

**Comment by Klap-in** [comment]
Regarding code style, see also #628

**Comment by fcastilloec** [comment]
Any updates on this PR?
It seems all work stop mainly because of styling issues. The irony is that https://github.com/pwr-Solaar/Solaar/pull/628#issuecomment-573425573 says: 

> I don't think we want to stall contributions because they don't follow the code style.

Which is sort of happening in this PR, this PR is not following the code style of this repo (**absence of a style is a style**).
Nonetheless, can @Vtec234 just force push new commits without any formatting changes? At the end of the day, the maintainers are asking to follow their style, and I do agree that it makes it harder for them to review the changes if they also have to review style changes.
If worst comes to It (@Vtec234 is too busy, or has lost interest, etc.), I could replicate this PR, rebased it over master, and make sure there aren't any style changes. I just want to make sure this amazing contribution gets it to the users.

**Comment by Klap-in** [comment]
Recently a code style is applied everywhere.  https://github.com/pwr-Solaar/Solaar/pull/835  Further there are also developments with the features changed here. So some work would be needed before merging is possible.

For further guiding @pfps can better respond here.

**Comment by pfps** [comment]
One problem with this PR is that at least one commit changed a lot of lines (from style changes) and also changed a few lines (from actual changes).  This meant that it was very hard to review and I never got around to a complete review.

I finally grabbed a couple of minor things out of some of the commits and put them in Solaar.

Since then some of the code and data structures that this PR changed have changed to support some of the new features of Solaar.  And, Solaar now has a code style, which is different from the style in this PR, so there is quite a bit of work that is needed to make this PR fit into the current version of Solaar.

There is also the issue that this PR adds an entirely new kind of interaction for Solaar.  There should be a determination of whether this is a good idea for Solaar.  I'll add a new comment here about this issue.

**Comment by pfps** [comment]
This PR makes Solaar intercept regular output of devices - mouse movements - to affect the state of the device - changing the DPI.  This is how it is done in the Logitech software but is this the right way for Solaar?

The alternative is to just change DPI using a setting in the Solaar main window.  This not as simple as with Logitech, of course.

**Comment by fcastilloec** [comment]
@pfps @Klap-in Thanks for the replies. Sorry for not noticing the styling PR, I was just going from what was mentioned in this PR. I'm glad to see that styling won't be an issue in the future!

>The alternative is to just change DPI using a setting in the Solaar main window. This not as simple as with Logitech, of course.

Isn't this already implemented? I'm able to change the DPI of the mouse from Solaar main window without a problem. This PR was trying to get the DPI button to work.

>This PR makes Solaar intercept regular output of devices - mouse movements - to affect the state of the device - changing the DPI. This is how it is done in the Logitech software but is this the right way for Solaar?

I do agree that this should be discussed in more detail. I do not like the way Logitech designed this (I would prefer using the scroll wheel rather than move movement) but on the other hand, it'll be good to have consistency for people coming from Windows/macOS and expecting the same behavior. Offering both behaviors seems like too much unnecessary work.

What hasn't been implemented in this PR is changing the DPI with a single click of a button. In https://github.com/pwr-Solaar/Solaar/pull/733#issuecomment-613560385 this is mentioned and hopefully, @Vtec234 has implemented something.
There could be a setting in the main window with two DPI values, and when the button is pressed, alternate between them.
From my point of view, this last feature is more important than changing DPI by mouse movement.

**Comment by pfps** [comment]
Yes, Solaar can set DPI so as far as I can tell this PR just adds a new way to set DPI.

**Comment by Vtec234** [comment]
Hello again!
@fcastilloec: I've indeed been quite busy, but have also been meaning to come back to this eventually. I actually have some free time this weekend which could be spent on this.

> Recently a code style is applied everywhere

I think this is great. I can now just `yapf` all my patches and not worry about formatting at all. It should speed up the process.

> Isn't this already implemented? I'm able to change the DPI of the mouse from Solaar main window without a problem. This PR was trying to get the DPI button to work.

Exactly, changing DPI in the settings is already possible. The point of this PR is to allow doing this quickly (for this toggling with a press of the DPI button suffices) and smoothly (for this sliding is needed), e.g. while working with image manipulation software. I'd imagine people with gaming mice would also want this, but dunno if Solaar supports the gaming use case.

> I finally grabbed a couple of minor things out of some of the commits and put them in Solaar.

(By the way, I think proper attribution in the git history is important, so when committing my code please preserve the authorship information. This is possible either using an interactive rebase to pick out whichever commits are needed or by simply adding a `Co-authored-by: ` line to the commit message. I don't particularly mind for small changes, but e.g. #822 would've been a tangible change losing autorship if merged.)

> Yes, Solaar can set DPI so as far as I can tell this PR just adds a new way to set DPI.

Sort of, in terms of user-facing functionality. But under the hood it implements the entire `REPROG_CONTROLS_V4` feature including diverting keys. I see part of this has been [added](https://github.com/pwr-Solaar/Solaar/commit/9beb13a3053bd595959304840f33fd4bfef1ede5) in the meantime, but it seems without divertable controls or reporting of XY positions.

What I could do is temporarily forget about MX Vertical and just reintroduce the wrapper classes which implement `REPROG_CONTROLS_V4` (more or less #822) in a self-contained PR. To adapt to new changes, I would restate the code in `_feature_reprogrammable_keys_choices` in terms of this. Then DPI sliding will be implementable on top of that, and it can be a separate PR. Does this sound reasonable @pfps?

**Comment by pfps** [comment]
I think that what I ended up using was the better reporting of errors.  I should have acknowledged where this came from.  And probably a better way to do it would have been to cherry pick one our your commits and then modifying the commit as necessary.

**Comment by Vtec234** [comment]
Good news, I reimplemented the whole thing (almost) as it behaves on Windows under Logitech's software. That is, there are two DPI slots which you can switch between with a DPI-button press. Holding the button down and sliding the mouse changes the DPI in the current slot. If you have any reasonable notification daemon installed, a progress bar should pop up while you do this visualizing what the current DPI is. @fcastilloec you might be interested

As you might expect of something I hacked together in a few hours, there are probably bugs and one missing bit is that the DPI slots' values are not persisted between runs of Solaar, so that when you restart they'll both be reinitialized to whatever the current DPI is.

It works fine for me so I'm not going to spend any more time on it, but the code quality is dubious. The whole thing holds together by duck tape and hacks around the Solaar design in a bunch of places, so in that respect I would actually advise *against* merging this. That said, users might want this feature, so there's a tradeoff. Ideally someone would rewrite the `Setting` API to be more flexible (arbitrary callbacks rather than "give me some bytes to send to the device").

**Comment by pfps** [comment]
On Windows, when the mouse is sending events that change the DPI does the mouse also move?

For me, DPI changing is very slow - about one slot / 10 cm.  I guess that there is something wrong in the slot per movement calculation.

It seems to me that the notification is being updated every time the mouse moves.  Wouldn't it be better to only notify if the calculated dpi changes?

**Comment by pfps** [comment]
A commit I made causes a conflict in this PR.  The fix is easy.  I tried to fix it up myself but somehow my fix didn't completely take.  I think what I needed to do was to push to the branch on @Vtec234's repository, which I probably can't do.

**Comment by Vtec234** [comment]
> On Windows, when the mouse is sending events that change the DPI does the mouse also move?

It doesn't, and shouldn't do that on Linux either -- the mouse firmware should block normal mouse movement events while xy redirection is enabled. Does the mouse move for you?

> For me, DPI changing is very slow - about one slot / 10 cm

That doesn't sound right, is this on a Vertical mouse or are you testing on something else? You could add a `print` [here](https://github.com/pwr-Solaar/Solaar/blob/27cb1a197fd8da2944edeb0e978119f4a9720194/lib/logitech_receiver/settings_templates.py#L436) and see what the value of `self.dx` is. It should be somewhere around 330 after 10cm.

> Wouldn't it be better to only notify if the calculated dpi changes?

Good idea.

**Comment by pfps** [comment]
Putting a print in the inner loop was slow.  What I did was print the resultant DPI.  The result was 9 slots for roughly the same motion that takes me across my entire screen.  Changing the multiplier to 500. results in a much better experience for me but maybe a value around.  Notifying only on DPI changes results in much better displaying (under XFCE).

**Comment by pfps** [comment]
I think a couple of things are needed to make this ready to merge:
- the DPI setting should be affected
- the check should be for the presence of a DPI button, not the MX vertical
- it should work when the MX vertical is connected via a USB cable.
The last should fall out from the second, as far as I can tell.

**Comment by fcastilloec** [comment]
> the check should be for the presence of a DPI button, not the MX vertical

Are we sure that all Logitech mice with a DPI button have the same interface for changing DPI? I would guess that's true, but it'll be a good idea to double-check, otherwise, users might be expecting a specific behavior and finding something else.

**Comment by pfps** [comment]
@fcastilloec The check would be for a button that has the capabilities needed - sending the DPI button message and sending raw mouse movements when depressed.  (It may be that only the MX Vertical has such a button.)  If such a button exists then its purpose is going to be for changing DPI.

**Comment by Vtec234** [comment]
> The result was 9 slots for roughly the same motion that takes me across my entire screen.

Ah but you didn't say if this is also on a Vertical mouse? I assumed that sensors in all mice of this model would be comparable, but if you're seeing such divergence then clearly this isn't the case, which is a bit annoying.

**Comment by pfps** [comment]
This is for an MX Vertical.     
```
2: Wireless Mouse MX Vertical
     Device path  : /dev/hidraw9
     WPID         : 407B
     Codename     : MX Vertical
     Kind         : mouse
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: FD48213A
     Model ID:      B020407BC08A
     Unit ID:       5EEE2183
        Bootloader: BOT 78.00.B0009
          Firmware: MPM 16.00.B0009
             Other: 
```

**Comment by pfps** [comment]
@Vtec234 Let me know if you are going to work on this more.  If not, I may be able to work on it.

**Comment by Vtec234** [comment]
@pfps I rebased on master and fixed the notification spam, so that now it only updates the notification if there actually is something to update, i.e. the selected DPI has changed. I leave the rest up to you. As a maintainer of Solaar you *should* have push access to this branch in my repository, but just in case that doesn't work I also sent a collaboration invite.

**Comment by pfps** [comment]
OK I made some changes to your branch.


The setting now updates the main window with the new value.

The setting is now supported by a device if the device has
the REPROG_CONTROLS_V4 feature, a DPI_Switch key that can be diverted to produce raw XY reports, and the ADJUSTABLE_DPI feature.



Give them a try.

**Comment by pfps** [comment]
There are a couple of spit-and-polish changes to think about.  
1/ How sensitive to make the slider.  I was too low in my previous estimate of the sensitivity (because I didn't wait for all the notifications to be processed) but the sensitivity is too low, I think.
2/ When to switch between a simple switch of sensitivity and a sliding switch.   Several times I clicked the DPI button but the sensitivity didn't change because the mouse moved.  Maybe the determination on which on to use can be whether the mouse moved enough to slide to a new sensitivity.

**Comment by pfps** [comment]
I adjusted the sensitivity of DPI adjustments to about 5/cm.
I delayed the switch from pressed to moved until there is enough movement to trigger a DPI change.

**Review by pfps — APPROVED** [review]
I've added some features to the PR and adjusted the behaviour a bit.
@Vtec234 Let me know if you think the modified sensitivity and change-over are acceptable.  If so, I'll merge this into the  main branch.

**Comment by Vtec234** [comment]
Yeah, they are an actual improvement, so I'm using the new version now. Thanks!

**Comment by pfps** [comment]
OK, good enough for me.

**Comment by fcastilloec** [comment]
Sorry for being so late in testing this PR. Just wanted to say that everything is working great! Sliding and single pressing are both working as intended.
I have a couple of comments, which I probably will open a new issue, but thought it would be a good idea to also put them here.

* When sliding, I see a percentage and nothing else. There should be some extra info to it, otherwise, people might confuse it with the battery level. Either add something to the percentage or show the actual DPI info.
* It seems that changes to the DPI from the settings are no longer respected after a single press of the DPI button. To be more specific, I've set two DPI values by sliding, which I can switch by a single press of the DPI button. If I change the DPI from the settings window and then press the DPI button twice (in order to toggle back to the same preset), the value goes back to the original one set by sliding, and not to the value set using the settings window. This happens with either of the two presets.
* If the mouse goes into power-savings or is switched off and on, neither of the DPI presets (set by sliding) are saved. Maybe there's a way to save the settings and reload them once the mouse is active/on.

If the people involved with this PR think the previous points are valid, I'll go ahead and open a new issue so we can track them.

**Comment by pfps** [comment]
@fcastilloec 

* The OSD for sliding is done on-the-cheap.  Something better would require significantly more effort.  That's not to say that it isn't a good idea, just that someone would have to do the work.
* Yes, clicking the DPI button uses only the sliding values.  This might be easy to fix.
* Losing the presets is an issue.  There might be a relatively easy fix for this.

**Comment by pfps** [comment]
@fcastilloec So please open an issue for your last two points.  Your first point is unlikely to be worked on, but you could open an issue for that one as well, but please make it a separate issue.

**Comment by pfps** [comment]
@fcastilloec Try out PR #962.  See whether it solves the last two of your issues.  Say whether you like the value notifications over the percentage notifications.

**Comment by pfps** [comment]
@Vtec234 I'm going to upgrade this setting a bit, turning it from a boolean to a choice, where the choices are Off and the DPI Switch button if there is one.  If there isn't one the choices will be Off and the buttons that can be diverted and cause raw XY reporting.   This will allow sliding DPI on more devices.
See the new mouse gestures setting for how this will work.

**Pull request merged**
